server/src/main/java/org/opensearch/index/engine/InternalEngine.java

```diff
@@ -221,58 +222,86 @@ public class InternalEngine extends Engine {
-        }
-        if (customTranslogDeletionPolicy != null) {
-            translogDeletionPolicy = customTranslogDeletionPolicy;
-        } else {
-            translogDeletionPolicy = new DefaultTranslogDeletionPolicy(
-                engineConfig.getIndexSettings().getTranslogRetentionSize().getBytes(),
-                engineConfig.getIndexSettings().getTranslogRetentionAge().getMillis(),
-                engineConfig.getIndexSettings().getTranslogRetentionTotalFiles()
-            );
-        }
+        final TranslogDeletionPolicy translogDeletionPolicy = getTranslogDeletionPolicy(engineConfig);
         store.incRef();
         IndexWriter writer = null;
-        Translog translog = null;
         ExternalReaderManager externalReaderManager = null;
         OpenSearchReaderManager internalReaderManager = null;
         EngineMergeScheduler scheduler = null;
+        TranslogManager translogManagerRef = null;
         boolean success = false;
         try {
             this.lastDeleteVersionPruneTimeMSec = engineConfig.getThreadPool().relativeTimeInMillis();
             mergeScheduler = scheduler = new EngineMergeScheduler(engineConfig.getShardId(), engineConfig.getIndexSettings());
             throttle = new IndexThrottle();
             try {
                 store.trimUnsafeCommits(engineConfig.getTranslogConfig().getTranslogPath());
-                translog = openTranslog(engineConfig, translogDeletionPolicy, engineConfig.getGlobalCheckpointSupplier(), seqNo -> {
-                    final LocalCheckpointTracker tracker = getLocalCheckpointTracker();
-                    assert tracker != null || getTranslog().isOpen() == false;
-                    if (tracker != null) {
-                        tracker.markSeqNoAsPersisted(seqNo);
+                final Map<String, String> userData = store.readLastCommittedSegmentsInfo().getUserData();
+                final String translogUUID = Objects.requireNonNull(userData.get(Translog.TRANSLOG_UUID_KEY));
+                TranslogEventListener internalTranslogEventListener = new TranslogEventListener() {
+                    @Override
+                    public void onAfterTranslogSync() {
+                        revisitIndexDeletionPolicyOnTranslogSynced();
                     }
-                });
-                assert translog.getGeneration() != null;
-                this.translog = translog;
+
+                    @Override
+                    public void onAfterTranslogRecovery() {
+                        flush(false, true);
+                        translogManager.trimUnreferencedTranslogFiles();
+                    }
+
+                    @Override
+                    public void onFailure(String reason, Exception ex) {
+                        failEngine(reason, ex);
+                    }
+
+                    @Override
+                    public void onTragicFailure(AlreadyClosedException ex) {
+                        failOnTragicEvent(ex);
```

Comment:
this is a nit - I'm wondering if we should leave it up to the engine to determine if the error is tragic or not instead of explicitly catching `AlreadyClosedException` in `InternalTranslogManager`?  Then we don't need this separation in `TranslogEventListener`.